pkg/controller/crdmirroring/crdmirroring_controller.go

```diff
@@ -0,0 +1,781 @@
+
+		legacyInformer := legacyGeneralInformer.(legacycoreinformer.ExternalEntityInformer)
+		legacyHandlers := cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onLegacyExternalEntityAdd,
+			UpdateFunc: c.onLegacyExternalEntityUpdate,
+			DeleteFunc: c.onLegacyExternalEntityDelete,
+		}
+		legacyInformer.Informer().AddEventHandler(legacyHandlers)
+		c.legacyLister = legacyInformer.Lister()
+		c.legacyListerSynced = legacyInformer.Informer().HasSynced
+
+	case coreinformer.ClusterGroupInformer:
+		informer := generalInformer.(coreinformer.ClusterGroupInformer)
+		handlers := cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onClusterGroupAdd,
+			UpdateFunc: c.onClusterGroupUpdate,
+			DeleteFunc: c.onClusterGroupDelete,
+		}
+		informer.Informer().AddEventHandler(handlers)
+		c.lister = informer.Lister()
+		c.listerSycned = informer.Informer().HasSynced
+
+		legacyInformer := legacyGeneralInformer.(legacycoreinformer.ClusterGroupInformer)
+		legacyHandlers := cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onLegacyClusterGroupAdd,
+			UpdateFunc: c.onLegacyClusterGroupUpdate,
+			DeleteFunc: c.onLegacyClusterGroupDelete,
+		}
+		legacyInformer.Informer().AddEventHandler(legacyHandlers)
+		c.legacyLister = legacyInformer.Lister()
+		c.legacyListerSynced = legacyInformer.Informer().HasSynced
+	}
+
+	c.reconciler = newReconciler(CRDClient, legacyCRDClient)
+	c.tracker = newTracker()
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %vMirroringController", c.CRDName)
+	defer klog.Infof("Shutting down %vMirroringController", c.CRDName)
+
+	if !cache.WaitForNamedCacheSync(fmt.Sprintf("%vMirroringController", c.CRDName), stopCh, c.listerSycned, c.legacyListerSynced) {
+		return
+	}
+
+	klog.Infof("Starting %d worker threads", defaultWorkers)
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncCRD(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) queueEndpoints(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v (type %T): %v", obj, obj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *Controller) syncCRD(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing for %q legacy CRD. (%v)", key, time.Since(startTime))
+	}()
+
+	klog.V(4).Infof("Sync legacyClient CRD (%q)", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	switch c.legacyLister.(type) {
+	case legacysecuritylister.NetworkPolicyLister:
```

Comment:
Using many this kind of switches may be not different from rewriting each controller directly. Let's consider if there is more elegant way.